db/snapshot/backup_task.cc

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ *
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "utils/lister.hh"
+#include "utils/s3/client.hh"
+#include "replica/database.hh"
+#include "db/config.hh"
+#include "db/snapshot-ctl.hh"
+#include "db/snapshot/backup_task.hh"
+#include "schema/schema_fwd.hh"
+#include "sstables/sstables.hh"
+
+extern logging::logger snap_log;
+
+namespace db::snapshot {
+
+backup_task_impl::backup_task_impl(tasks::task_manager::module_ptr module, snapshot_ctl& ctl, shared_ptr<s3::client> client, sstring bucket, sstring ks, sstring snapshot_name) noexcept
+    : tasks::task_manager::task::impl(module, tasks::task_id::create_random_id(), 0, "node", ks, "", "", tasks::task_id::create_null_id())
+    , _snap_ctl(ctl)
+    , _client(std::move(client))
+    , _bucket(std::move(bucket))
+    , _ks(std::move(ks))
+    , _snapshot_name(std::move(snapshot_name))
+{}
+
+std::string backup_task_impl::type() const {
+    return "backup";
+}
+
+tasks::is_internal backup_task_impl::is_internal() const noexcept {
+    return tasks::is_internal::no;
+}
+
+static std::pair<sstring, table_id> extract_cf_name_and_uuid(const sstring& directory_name) {
+    // cf directory is of the form: 'cf_name-uuid'
+    // uuid is assumed to be exactly 32 hex characters wide.
+    constexpr size_t uuid_size = 32;
+    ssize_t pos = directory_name.size() - uuid_size - 1;
+    if (pos <= 0 || directory_name[pos] != '-') {
+        on_internal_error(snap_log, format("table directory entry name '{}' is invalid: no '-' separator found at pos {}", directory_name, pos));
+    }
+    return std::make_pair(directory_name.substr(0, pos), table_id(utils::UUID(directory_name.substr(pos + 1))));
+}
+
+void backup_task_impl::run_in_thread(sstring data_dir) {
+    std::exception_ptr ex;
+    //
+    // The keyspace data directories and their snapshots are arranged as follows:
+    //
+    //  <data dir>
+    //  |- <keyspace name1>
+    //  |  |- <column family name1>
+    //  |     |- snapshots
+    //  |        |- <snapshot name1>
+    //  |          |- <snapshot file1>
+    //  |          |- <snapshot file2>
+    //  |          |- ...
+    //  |        |- <snapshot name2>
+    //  |        |- ...
+    //  |  |- <column family name2>
+    //  |  |- ...
+    //  |- <keyspace name2>
+    //  |- ...
+    //
+    auto ks_dir = fs::path(data_dir) / _ks;
+    auto ks_dir_lister = directory_lister(ks_dir, lister::dir_entry_types::of<directory_entry_type::directory>());
+    auto close_ks_dir_lister = deferred_close(ks_dir_lister);
+
+    snap_log.trace("Scan keyspace dir {}", ks_dir.native());
+    while (auto table_ent = ks_dir_lister.get().get()) {
+        auto snapshot_dir = ks_dir/ table_ent->name / sstables::snapshots_dir / _snapshot_name;
```

Comment:
nit: space before `/` in `ks_dir/`

Reply:
see also https://github.com/scylladb/scylladb/pull/19890#discussion_r1693138628

Reply:
Yup, double catch :) Fixed